lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1413,6 +1413,48 @@ RequirementSignatureRequest::evaluate(Evaluator &evaluator, ProtocolDecl *proto)
   return reqSignature->getRequirements();
 }
 
+llvm::Expected<Type>
+DefaultDefinitionTypeRequest::evaluate(Evaluator &evaluator,
+                                       AssociatedTypeDecl *assocType) const {
+  if (assocType->Resolver) {
+    auto defaultType = assocType->Resolver->loadAssociatedTypeDefault(
+                                    assocType, assocType->ResolverContextData);
+    assocType->Resolver = nullptr;
+    assocType->DefaultDefinition.setType(defaultType);
+    return defaultType;
+  }
+
+  auto &context = assocType->getASTContext();
+  auto &TC = TypeChecker::createForContext(assocType->getASTContext());
+
+  TypeLoc &defaultDefinition = assocType->DefaultDefinition;
+  if (!defaultDefinition.isNull()) {
+    if (TC.validateType(defaultDefinition,
```

Comment:
I think we should go further and rip out this TypeLoc entirely, replacing it with a TypeRepr stored in the AssociatedTypeDecl. Then instead of creating a TypeChecker and calling validateType(), you use resolveType() which does not need a TypeChecker:

```
auto resolution = TypeResolution::forInterface(cType->getDeclContext());
auto type = resolution.resolveType(...);
```

This will let you use the request evaluator's own caching too.

Reply:
Good idea, this change cleaned up some complex code.